File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        
        int n = nums.size();
        
        if ( n == 1 ) return nums[0];
        
        int r1 = robber( 0, n - 2, nums );
        int r2 = robber( 1, n - 1, nums );
        
        return max( r1, r2 );
        
    }
    
    int robber( int start, int end, const vector<int>& nums ) {
        
        int prev = 0;
        int curr = 0;
        int next = 0;
        
        for (int i = start; i <= end; i++) {
            
            next = max( curr, prev + nums[i] );
            prev = curr;
            curr = next;
        }
        
        return curr;
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]

        # A helper function to do the robbing for a linear street of houses
        def rob_linear(houses):
            prev1, prev2 = 0, 0
            for money in houses:
                temp = prev1
                prev1 = max(prev1, prev2 + money)
                prev2 = temp
            return prev1

        # Case 1: Rob houses from 0 to n-2
        case1 = rob_linear(nums[:-1])
        # Case 2: Rob houses from 1 to n-1
        case2 = rob_linear(nums[1:])

        # Return the maximum of the two cases
        return max(case1, case2)
        